[FEAT]  회원가입 페이지 폼 예외처리 기능 구현

kimyou1102 opened the pull request:
## Issue Number
closed #202

## As-Is
<!-- 문제 상황 정의 -->
- 회원가입 입력 검증이 없음

## To-Be
<!-- 변경 사항 -->
- 이름 길이 및 입력값 검증
- 아이디 길이 및 입력값 검증
- 비밀번호 길이 검증
- 비밀번호와 비밀번호 일치 검증
- 전화번호 검증
- 인증번호 확인 길이 검증
- 모든 값이 유효할 경우만 제출되도록 구현
- 검증 훅 테스트 작성

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description
- 각 검증은 중요 로직이라 훅테스트로 작성하고, 버튼의 활성화 여부는 css요소가 있기 때문에 컴포넌트 테스트로 진행했어!
  - 지금은 현재 페이지가 넘어가거나 제출후의 상황을 구현하지 않아서 css 요소로만 검증했어
- 버튼의 활성화 여부를 결정하려면 각 입력값을 실시간으로 알아야 해서 SignupForm 컴포넌트에 모든 값이 모이는 구조가 됐는데 더 좋은 방법 있으면 공유해주면 고마울 거 같아요~~~🙇‍♀️

kimyou1102 committed:
style: 공통 Button 컴포넌트 pointer-events: none 속성 추가 #202

- disabled 상태에서도 cursor만 변경되고 실제 클릭은 막히지 않는 이슈 발생
- 클릭 자체를 방지하기 위해 pointer-events: none 속성 추가

kimyou1102 committed:
refactor: useMenteeNameInput 훅 공통훅으로 변경 #202

- 회원가입 페이지의 이름입력 시 사용하는 로직과 공통되어서 공통 훅으로 변경
- 공통훅으로 사용하기 위해 useMenteeNameInput -> useNameInput 네이밍 변경

kimyou1102 committed:
refactor: 테스트 파일명 변경 #202

kimyou1102 committed:
refactor: name 상수 최소값 변경 #202

kimyou1102 committed:
refactor: 이름 길이 비교 name상수 사용 #202

kimyou1102 committed:
feat: 이름 검증 훅 이용해서 구현 #202

kimyou1102 committed:
feat: 모든 입력폼이 유효하지 않으면 제출 버튼을 disabled 처리 #202

kimyou1102 committed:
feat: 성별 선택 라디오버튼 value state 연결 #202

kimyou1102 committed:
feat: validateLength 유틸함수 생성 #202

kimyou1102 committed:
feat: usePassword 훅 생성 #202

kimyou1102 committed:
feat: 비밀번호 검증 구현 #202

kimyou1102 committed:
style: 스타일 속성 순서 변경 #202

kimyou1102 committed:
refactor: usePassword ->  usePasswordInput 훅이름 변경 #202

kimyou1102 committed:
feat: useUserIdInput 아이디 검증 훅 생성 #202

kimyou1102 committed:
feat: useUserIdInput 훅 사용해서 아이디 검증 기능 구현 #202

kimyou1102 committed:
refactor: getPhoneNumberErrorMessage 유틸함수, useFormattedPhoneNumber 훅 common으로 이동

kimyou1102 committed:
feat: 전화번호 포매팅 및 검증기능 구현 #202

kimyou1102 committed:
feat: useVerificationCodeInput 인증 번호 입력 검증 훅 구현 #202

kimyou1102 committed:
feat: 인증번호 확인 검증 기능 구현 #202

kimyou1102 committed:
fix: getPasswordErrorMessage 조건문 수정 #202

- passwordConfrim의 값도 입력시에 검증하게 되어서 오류 발생
- 불필요한 passwordConfrim 입력값 체크 제거

kimyou1102 committed:
feat: 인풋 훅들 isValid 리턴값 추가 #202

- 해당 훅이 유효한지에 대한 값을 훅에 포함하면 재사용될 여지가 높기에 추가

kimyou1102 committed:
feat:모든 입력폼이 유효한지 검증하는  isFormValid 함수 구현 #202

-  리턴값만 있던 isFormValid함수 내용 구현

kimyou1102 committed:
fix: validateLength유틸함수 조건문 변경 #202

- 미만 초과가 아니라  이상 이하값을 반환하기 위해 조건문 변경

kimyou1102 committed:
test: usePasswordInput 훅 테스트 작성 #202

kimyou1102 committed:
test: useUserIdInput 훅 테스트 작성 #202

kimyou1102 committed:
test: useVerificationCodeInput 훅 테스트 작성 #202

kimyou1102 committed:
refactor: 불필요한 콘솔로그 제거 #202

kimyou1102 committed:
test: SignupForm 버튼 제출 버튼 활성화 여부 테스트 작성 #202

Beomtae commented on frontend/__tests__/SignupForm.test.tsx:
```diff
@@ -0,0 +1,68 @@
+
+    const submitButton = screen.getByRole('button', { name: /회원가입/i });
+
+    expect(submitButton).not.toHaveStyle('pointer-events: none');
```
클릭 여부를 pointer-events 로 판단할 수 있구나!! 👍👍👍

dlsxjzld replied:
한 수 배워갑니다~~👍

Beomtae commented on frontend/src/common/hooks/useFormattedPhoneNumber.ts:
```diff
@@ -1,7 +1,7 @@
 import { useRef, useState } from 'react';
 
-import { getFormattedPhoneNumber } from '../utils/getFormattedPhoneNumber ';
+import { getFormattedPhoneNumber } from '../../pages/booking/utils/getFormattedPhoneNumber ';
```
p3: 의존성 방향이 반대로 흐르고 있는 것 같아 (DIP). 이렇게 되면 예약 페이지에 종속된 유틸을 다른 컴포넌트(예: MyPage, 회원가입, 프로필) 등에서도 가져다 쓰는 형태가 될 수 도 있을 것 같은데 getFormattedPhoneNumber 도 common으로 옮기는 것에 대해 어떻게 생각해?

kimyou1102 replied:
공통된 부분이라 뺀다는걸 잊었네😅  완전 좋은 의견같아!! 반영할게👍👍

kimyou1102 replied:
수정커밋: [refactor: 공통적으로 사용되는 유틸 및 상수를 common 디렉토리로 이동](https://github.com/woowacourse-teams/2025-Fit-toring/pull/245/commits/92ed3b20cbad27b44110276e7ac6621b3d91c0dd)

Beomtae commented on frontend/src/common/hooks/useNameInput.ts:
```diff
@@ -0,0 +1,41 @@
+import { useState } from 'react';
+
+import { ERROR_MESSAGE } from '../../pages/booking/constants/errorMessage';
+import { NAME } from '../../pages/booking/constants/name';
```
이것도 마찬가지로 의존성 역전인 것 같아!

kimyou1102 replied:
수정커밋: [refactor: 공통적으로 사용되는 유틸 및 상수를 common 디렉토리로 이동](https://github.com/woowacourse-teams/2025-Fit-toring/pull/245/commits/92ed3b20cbad27b44110276e7ac6621b3d91c0dd)

Beomtae commented on frontend/src/common/utils/phoneNumberValidator.ts:
```diff
@@ -1,4 +1,4 @@
-import { ERROR_MESSAGE } from '../constants/errorMessage';
+import { ERROR_MESSAGE } from '../../pages/booking/constants/errorMessage';
```
이것도 마찬가지로 의존성 역전인 것 같아!

kimyou1102 replied:
수정커밋: [refactor: 공통적으로 사용되는 유틸 및 상수를 common 디렉토리로 이동](https://github.com/woowacourse-teams/2025-Fit-toring/pull/245/commits/92ed3b20cbad27b44110276e7ac6621b3d91c0dd)

Beomtae commented on frontend/src/pages/signup/hooks/useUserIdInput.ts:
```diff
@@ -0,0 +1,48 @@
+import { ID } from '../constants/id';
+
+const useUserIdInput = () => {
+  const [userId, setUserIdId] = useState('');
```
p3: setUserIdId 오타 난 것 같아

kimyou1102 replied:
헉 맞네 수정할게!!

kimyou1102 replied:
수정커밋: [refactor: setState 함수명 오타 수정](https://github.com/woowacourse-teams/2025-Fit-toring/pull/245/commits/dfb3bc777a2c3b124b5faab37dca2c49d09c32dd)

Beomtae reviewed (commented):
SignupForm 컴포넌트에 모든 값이 모이는 구조에 대해 더 좋은 방법이 있는지에 대해서는 나도 조금 더 생각을 해봐야 할 것 같아!! 예외처리 꼼꼼하게 너무 잘 해줬고 의존성 역전에 대해 코멘트 조금 달아놨어! 보고 코멘트 남겨줘 고생했어!!

kimyou1102 committed:
refactor: 공통적으로 사용되는 유틸 및 상수를 common 디렉토리로 이동

- 상위 모듈인 공통 훅에서 하위 모듈인 booking 유틸/상수에 의존하는 구조였음
- 의존성 방향이 거꾸로 설정된 문제를 해결하고, 재사용성을 높이기 위해 common으로 분리

kimyou1102 committed:
refactor: setState 함수명 오타 수정 #202

kimyou1102 committed:
refactor: 변경된 경로에 맞게 임포트문 경로 수정 #202

dlsxjzld commented on frontend/__tests__/SignupForm.test.tsx:
입력 폼이 제출이 되는 경우와 안 되는 경우를 잘 나눠서 보기 좋았어 👍

현재 입력 폼이 제출 안 되는 경우가 **"모든 입력값이 유효하지 않으면"** 만 있는데 
**"입력값이 하나라도 유효하지 않으면 버튼이 클릭되지 않는다."** 라는 테스트 케이스도 추가되면 어떨까? 🤔

당연히 입력들 중에 하나라도 유효하지 않으면 안 될 것을 알고 있지만, 시간이 지나면 잊어버릴 것 같아서 말해봤어 ㅎㅎ

이름, 아이디, 비밀번호 등 다양한 입력이 있는데 이 모든 걸 테스트하자는 건 아니고,
다른 값들은 다 정상인데 이름만 안되는 테스트 케이스를 추가하는 거 어때? 🤔

kimyou1102 replied:
유효하지 않은것에 대한 테스트도 추가하면 좋을거 같네! 반영할게~~

kimyou1102 replied:
수정커밋: [test: 회원가입 폼에서 하나의 인풋이라도 입력값이 유효하지 않을시 버튼 비활성화 테스트 추가](https://github.com/woowacourse-teams/2025-Fit-toring/pull/245/commits/f278f287b1933a2d79a479981d5d6713c756f2dc)

dlsxjzld commented on frontend/src/common/components/Button/Button.tsx:
```diff
@@ -79,5 +79,5 @@ const disabledStyles = (theme: myTheme) => css`
 
   color: ${theme.FONT.G01};
-  cursor: not-allowed;
+  pointer-events: none;
```
p4: pointer-events로 아예 버튼이 안 눌리게 하는 점 너무 좋다 👍

현재는 버튼 위에 마우스를 올려도 아무 것도 나오지 않아서 누를 수 없다는 버튼인지 색상으로만 구분 가능할 것 같아!

cursor: not-allowed가 적용되면 좀 더 사용자에게 명확해서 좋을 것 같은데 어떻게 생각해? 🧐

```tsx
<StyledButtonWrapper variant={variant} aria-label="버튼">
  <StyledContainer
      customStyle={customStyle}
      variant={variant}
      size={size}
      onClick={onClick}
    >
      {children}
    </StyledContainer>
<StyledButtonWrapper>

....

const StyledButtonWrapper = styled.div<{ variant: Variant }>`
  width: fit-content; // button 크기에 맞추기 위해 필요
  cursor: ${({ variant }) =>
    variant === 'disabled' ? 'not-allowed' : 'pointer'};
`;

```

kimyou1102 replied:
두가지 속성을 모두 사용하니 cursor가 속성이 안먹히는걸 봐서 불필요한 속성같아서 제거했어!

dlsxjzld replied:
오 맞아 두 가지 속성을 모두 적용하면 **pointer-events가 마우스로 적용되는 모든 걸 막아서** 안먹힌다고 하더라고 ㅎㅎ

그래서 StyledButtonWrapper라는 div 태그로 StyledContainer(button 태그)를 감싸고, 여기에 cursor 속성을 두면 적용이 되더라고!

**누를 수 없는 버튼인지 구별 할 수 있는 정보가 색상만 있어서**, 
마우스를 누를 수 없는 버튼 위에 올렸을 때 cursor가 not-allowed 인 정보가 추가로 보이면 좋을 거 같다는 의견이였어! 😊

kimyou1102 replied:
지금과 같은 구조에서는 button에 customsStyle을 주는 경우라 부모를 두면 flex-grow같은 속성이 안먹더라고
이걸 해결하기 위해 부모인 div에 customStyle을 주고 버튼의 속성을 모두 지우는 것도 함부로 하면 안될거 같아서 일단 보류해 두고 우리끼리 회의후에 정하는 거로 하자!!

dlsxjzld commented on frontend/src/common/constants/apiEndpoints.ts:
상수화 👍

dlsxjzld commented on frontend/__tests__/usePasswordInput.test.ts:
```diff
@@ -0,0 +1,101 @@
+import { ERROR_MESSAGE } from '../src/pages/signup/constants/errorMessage';
+import usePasswordInput from '../src/pages/signup/hooks/usePasswordInput';
+
+const createFakeChangeEvent = (
+  value: string,
+): React.ChangeEvent<HTMLInputElement> => {
+  return {
+    target: { value },
+  } as React.ChangeEvent<HTMLInputElement>;
+};
```
오 이렇게 fakeChangeEvent를 만들어서 넘기는구나! 
배워갑니다 👍👍

kimyou1102 replied:
renderHook 환경에서는 DOM이 없기 때문에 target.value를 가진 가짜 이벤트 객체를 만들어 onChange에 넘겨줬어

이 객체를 ChangeEvent로 인식하게 하기 위해 as로 단언이 필요했고 이걸 반복적으로 사용하게 되니 함수로 분리한 거야!!

dlsxjzld commented on frontend/__tests__/usePasswordInput.test.ts:
```diff
@@ -0,0 +1,101 @@
+      );
+    });
+
+    it('비밀번호가 5자일 때는 유효하다', () => {
```
경계값 테스트 👍👍

dlsxjzld commented on frontend/__tests__/usePasswordInput.test.ts:
```diff
@@ -0,0 +1,101 @@
+      );
+    });
+
+    it('비밀번호와 비밀번호 확인이 일치하면 에러 메시지를 반환하지 않는다', () => {
```
p5: 반환하지 않는다 에서 반환값이 없을 것 같은 느낌이라, "비밀번호와 비밀번호 확인이 일치하면 에러메시지로 빈 문자열을 반환한다" 는 어때?? 🤔

kimyou1102 replied:
더 정확한 것은 빈문자열이겠네 좋은의견 굿굿👍

kimyou1102 replied:
수정커밋: [test: 테스트 설명 더 정확하게 변경](https://github.com/woowacourse-teams/2025-Fit-toring/pull/245/commits/32ed92b67a97fd197f1f5ae1bdcc84208abc305d)

dlsxjzld commented on frontend/__tests__/useUserIdInput.test.ts:
```diff
@@ -0,0 +1,79 @@
+  });
+
+  describe('아이디 길이 유효성 검사', () => {
+    it(`길이가 ${ID.MIN_LENGTH - 1}자 이하면 에러 메시지를 반환한다`, () => {
```
p5: 여기는 상수를 쓴 이유가 있을까??

kimyou1102 replied:
유효한 아이디 길이가 변경될 수도 있을거 같아서 상수를 사용했어!

dlsxjzld commented on frontend/src/pages/signup/components/PasswordFields/PasswordFields.tsx:
의미론적으로 묶인 거 너무 좋은데요~ 👍

dlsxjzld commented on frontend/src/pages/signup/components/SignupForm/SignupForm.tsx:
```diff
@@ -1,27 +1,118 @@
+    isValid: isVerificationCodeValid,
+  } = useVerificationCodeInput();
+
+  const isFormValid = () => {
+    const validations = [
+      isNameValid,
+      isUserIdValid,
+      isPasswordValid,
+      isPasswordConfrimValid,
+      phoneNumber !== '' && phoneNumberErrorMessage === '',
+      isVerificationCodeValid,
+    ];
```
p3: valid가 형용사여서 isValid로 하기로 했던 거 같아서, `isFormValid`대신 `validateForm`은 어떨까? 🤔

 isFormValid는 is가 붙은 함수인데, validations 안에 있는 is가 붙은 값들은 다 boolean 형태인 거 같아서 헷갈릴 수 있을 거 같아!

kimyou1102 replied:
오 그게 더 직관적일거 같다 수정할게!!

kimyou1102 replied:
수정커밋: [refactor: 직관적인 이름으로 함수명 변경](https://github.com/woowacourse-teams/2025-Fit-toring/pull/245/commits/51d693e5443cfd80cb3c440517f312c4a7301cae)

dlsxjzld commented on frontend/src/pages/signup/hooks/usePasswordInput.ts:
```diff
@@ -0,0 +1,67 @@
+    handlePasswordChange,
+    handlePasswordConfirmChange,
+    passwordErrorMessage,
+    passwordConfirmErrorMessage: getPasswordConfirmError(),
```
p4: 위에 변수를 만들어놔서 이렇게 수정해도 될 거 같아 👍
```suggestion
    passwordConfirmErrorMessage
```

kimyou1102 replied:
수정커밋: [refactor: 만들어둔 상수화 사용하도록 수정](https://github.com/woowacourse-teams/2025-Fit-toring/pull/245/commits/68fad6d414d15b714b41232b865de6806b10580f)

dlsxjzld reviewed (commented):
form 예외처리 기능 구현하느라 고생 많았어!!
테스트도 정말 잘 작성했고, form도 중앙 집중 형식으로 잘 만든 것 같아 👍👍

더 좋은 방법이 있는지 나도 고민해볼게! 😊

kimyou1102 committed:
test: 회원가입 폼에서 하나의 인풋이라도 입력값이 유효하지 않을시 버튼 비활성화 테스트 추가 #202

kimyou1102 committed:
test: 테스트 설명 더 정확하게 변경 #202

kimyou1102 committed:
refactor: 직관적인 이름으로 함수명 변경 #202

kimyou1102 committed:
refactor: 만들어둔 상수화 사용하도록 수정 #202
